### PR TITLE
[KOGITO-4299] Implement openapi endpoints

### PR DESCRIPTION
**Pull request by gitgabrio**
@danielezonca @evacchi  @jiripetrlik

Needed by https://github.com/kiegroup/kogito-examples/pull/564

see https://issues.redhat.com/browse/KOGITO-4299

### CODE REVIEW

File: drools/kogito-pmml-openapi/pom.xml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-params</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.commons</groupId>
+      <artifactId>commons-lang3</artifactId>
+      <version>3.8.1</version>
```
Comment (jiripetrlik) [inline drools/kogito-pmml-openapi/pom.xml:82]: Would it be possible to move this version somewhere up?

File: drools/kogito-pmml-openapi/src/main/java/org/kie/kogito/pmml/openapi/PMMLOASUtils.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+import static org.kie.kogito.pmml.openapi.api.PMMLOASResult.NUMBER;
+import static org.kie.kogito.pmml.openapi.api.PMMLOASResult.OBJECT;
+import static org.kie.kogito.pmml.openapi.api.PMMLOASResult.PROPERTIES;
+import static org.kie.kogito.pmml.openapi.api.PMMLOASResult.STRING;
+import static org.kie.kogito.pmml.openapi.api.PMMLOASResult.TYPE;
+
+public class PMMLOASUtils {
+
+    public static final String INFINITY_SYMBOL = new String(Character.toString('\u221E').getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8);
+
```
Comment (jiripetrlik) [inline drools/kogito-pmml-openapi/src/main/java/org/kie/kogito/pmml/openapi/PMMLOASUtils.java:59]: Please remove one new line.

File: drools/kogito-pmml-openapi/src/main/java/org/kie/kogito/pmml/openapi/PMMLOASUtils.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    public static boolean isPredicted(MiningField toVerify) {
+        return FIELD_USAGE_TYPE.PREDICTED.equals(toVerify.getUsageType()) ||
+                FIELD_USAGE_TYPE.TARGET.equals(toVerify.getUsageType());
+    }
+
+    public static String getMappedType(DATA_TYPE toMap) {
+        switch (toMap) {
+            case DATE:
```
Comment (jiripetrlik) [inline drools/kogito-pmml-openapi/src/main/java/org/kie/kogito/pmml/openapi/PMMLOASUtils.java:79]: Personally I would prefer to have one return for each statement. It shows more clear what is happening. But it is only my personal opinion. Feel free to ignore it.

File: kogito-codegen-modules/kogito-codegen-predictions/src/test/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGeneratorTest.java [hunk 4]
```diff
@@ -153,36 +170,141 @@ void setPathValue() {
...
-        pmmlRestResourceGenerator.setPredictionModelName(template);
+        assertTrue(TEMPLATE.getFieldByName("MODEL_NAME").isPresent());
+        final FieldDeclaration modelName = TEMPLATE.getFieldByName("MODEL_NAME").get();
+        assertFalse(modelName.getVariable(0).getInitializer().isPresent());
+        pmmlRestResourceGenerator.setPredictionModelName(TEMPLATE);
         try {
-            assertEquals(KIE_PMML_MODEL.getName(), methodCallExpr.getArgument(0).asStringLiteralExpr().asString());
+            assertTrue(modelName.getVariable(0).getInitializer().isPresent());
+            assertEquals(KIE_PMML_MODEL.getName(), modelName.getVariable(0).getInitializer().get().asStringLiteralExpr().asString());
         } catch (Exception e) {
```
Comment (jiripetrlik) [inline kogito-codegen-modules/kogito-codegen-predictions/src/test/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGeneratorTest.java:180]: I would prefer to add throw to the method signature instead of catching the exception and calling fail.

**Review by jiripetrlik — COMMENTED** [review]
Hello @gitgabrio and thank you for PR. I have few comments/suggestions.

File: drools/kogito-pmml-openapi/src/main/java/org/kie/kogito/pmml/openapi/PMMLOASUtils.java (hunk 3 above)
Comment (gitgabrio) [reply drools/kogito-pmml-openapi/src/main/java/org/kie/kogito/pmml/openapi/PMMLOASUtils.java:78]: @jiripetrlik 
I understand your point, but at the same time I think this way to deal the switch makes clear that, for example, DATE, DATE_TIME and STRING are managed the same way - it is not just to save few lines. Hope to have made my point clear.

File: drools/kogito-pmml-openapi/src/main/java/org/kie/kogito/pmml/openapi/PMMLOASUtils.java (hunk 2 above)
Comment (gitgabrio) [reply drools/kogito-pmml-openapi/src/main/java/org/kie/kogito/pmml/openapi/PMMLOASUtils.java:59]: @jiripetrlik 
done

File: drools/kogito-pmml-openapi/pom.xml (hunk 1 above)
Comment (gitgabrio) [reply drools/kogito-pmml-openapi/pom.xml:91]: @jiripetrlik 
done

**Comment by gitgabrio** [comment]
Jenkins retest this

File: drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/AbstractPMMLRestResource.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        org.kie.api.pmml.PMML4Result pmml4Result = prediction.evaluateAll(prediction.newContext(variables));
+        return java.util.Collections.singletonMap(pmml4Result.getResultObjectName(), pmml4Result.getResultVariables().get(pmml4Result.getResultObjectName()));
+    }
+
+    protected PMML4Result descriptive(Application application, String modelName, Map<String, Object> variables) {
+        org.kie.kogito.prediction.PredictionModel prediction = application.get(org.kie.kogito.prediction.PredictionModels.class).getPredictionModel(modelName);
+        return prediction.evaluateAll(prediction.newContext(variables));
+    }
+
+    public static String getJsonErrorMessage(Exception e) {
```
Comment (danielezonca) [inline drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/AbstractPMMLRestResource.java:37]: Why this change? It seems a general rewrite of the exception not specific for PMML so I don't understand the reason. Moreover in the endpoints you are adding a general exception mapper for all `Exception` so it is just a general catch all. Can you please explain?

File: kogito-codegen-modules/kogito-codegen-predictions/src/main/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGenerator.java [hunk 6]
```diff
@@ -106,17 +122,107 @@ void setPathValue(ClassOrInterfaceDeclaration template) {
...
     }
+
+    Optional<MemberValuePair> getRefMemberValuePair(NodeList<AnnotationExpr> source, String annotationName, String schemaName) {
+        return getAnnotationExpr(source, annotationName)
+                .map(annExpr -> getMemberValuePairFromAnnotation(annExpr, CONTENT))
+                .filter(Optional::isPresent)
+                .map(content -> getMemberValuePairFromMemberValuePair(content.get(), SCHEMA))
+                .filter(Optional::isPresent)
+                .map(schema -> getNormalAnnotationExprFromMemberValuePair(schema.get(), schemaName))
+                .filter(Optional::isPresent).flatMap(schemaAnnotation -> getMemberValuePairFromAnnotation(schemaAnnotation.get(), REF));
```
Comment (danielezonca) [inline kogito-codegen-modules/kogito-codegen-predictions/src/main/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGenerator.java:195]: You can just apply `map` without the need of a filter
```suggestion
        return getAnnotationExpr(source, annotationName)
                .map(annExpr -> getMemberValuePairFromAnnotation(annExpr, CONTENT))
                .map(content -> getMemberValuePairFromMemberValuePair(content.get(), SCHEMA))
                .map(schema -> getNormalAnnotationExprFromMemberValuePair(schema.get(), schemaName))
                .flatMap(schemaAnnotation -> getMemberValuePairFromAnnotation(schemaAnnotation.get(), REF));
```

File: kogito-codegen-modules/kogito-codegen-predictions/src/main/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGenerator.java [hunk 7]
```diff
@@ -106,17 +122,107 @@ void setPathValue(ClassOrInterfaceDeclaration template) {
...
+        ref = getRefMemberValuePair(annotations, QUARKUS_API_RESPONSE, QUARKUS_SCHEMA);
+        ref.ifPresent(rf -> rf.setValue(new StringLiteralExpr(outputRef)));
+    }
+
+    void setSpringOASAnnotations(NodeList<AnnotationExpr> annotations, String inputRef, String outputRef) {
+        Optional<MemberValuePair> ref = getRefMemberValuePair(annotations, SPRING_REQUEST_BODY, SPRING_SCHEMA);
+        ref.ifPresent(rf -> rf.setValue(new StringLiteralExpr(inputRef)));
+        ref = getRefMemberValuePair(annotations, SPRING_API_RESPONSE, SPRING_SCHEMA);
+        ref.ifPresent(rf -> rf.setValue(new StringLiteralExpr(outputRef)));
+    }
```
Comment (danielezonca) [inline kogito-codegen-modules/kogito-codegen-predictions/src/main/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGenerator.java:182]: I think both these methods can be moved to `DependencyInjectionAnnotator` interface :thinking: 
Let's consider this with a different ticket

File: drools/kogito-pmml-openapi/pom.xml [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>kogito-pmml-openapi</artifactId>
+
+  <name>Kogito :: Drools :: Kogito Predictive Model Markup Language :: OpenAPI</name>
+  <description>Kogito OpenAPI generator for Kie Drools PMML open source engine</description>
+
+  <properties>
+    <maven.compiler.source>11</maven.compiler.source>
+    <maven.compiler.target>11</maven.compiler.target>
```
Comment (danielezonca) [inline drools/kogito-pmml-openapi/pom.xml:19]: To be removed

File: drools/kogito-pmml-openapi/pom.xml [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    <version.smallrye-open-api-core>2.0.10</version.smallrye-open-api-core>
+    <version.commons-lang3>3.8.1</version.commons-lang3>
```
Comment (danielezonca) [inline drools/kogito-pmml-openapi/pom.xml:21]: Please move to [`kogito-build-parent`](https://github.com/kiegroup/kogito-runtimes/blob/master/kogito-build-parent/pom.xml) (also dependencyManagement section)

File: drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/AbstractPMMLRestResource.java (hunk 5 above)
Comment (gitgabrio) [reply drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/AbstractPMMLRestResource.java:37]: this returns a json-formatted message, that is what is required by the endpoint - otherwise the endpoint returns an exception because the body (that should be the original exception) is not json-formatted

File: kogito-codegen-modules/kogito-codegen-predictions/src/main/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGenerator.java (hunk 7 above)
Comment (gitgabrio) [reply kogito-codegen-modules/kogito-codegen-predictions/src/main/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGenerator.java:182]: @danielezonca 
That code is written  in terms of reusability whenever I find something that more or less evidently has/should be reused.
That class, and also the AbstractPmmlRestResource, should be extended/reused by other models IMO, but this goes beyond the current ticket

File: kogito-codegen-modules/kogito-codegen-predictions/src/main/resources/class-templates/PMMLRestResourceSpringTemplate.java [hunk 10]
```diff
@@ -3,24 +3,40 @@
...
-        PredictionModel prediction = application.get(PredictionModels.class).getPredictionModel("$modelName$");
-        return prediction.evaluateAll(prediction.newContext(variables));
+    @PostMapping(value = "/", produces = MediaType.APPLICATION_JSON_VALUE, consumes = MediaType.APPLICATION_JSON_VALUE)
+    @io.swagger.v3.oas.annotations.parameters.RequestBody(content = @io.swagger.v3.oas.annotations.media.Content(mediaType = "application/json",schema = @io.swagger.v3.oas.annotations.media.Schema(ref = "/pmmlDefinitions.json#/definitions/InputSet")), description = "PMML input")
+    @io.swagger.v3.oas.annotations.responses.ApiResponse(content = @io.swagger.v3.oas.annotations.media.Content(mediaType = "application/json",schema = @io.swagger.v3.oas.annotations.media.Schema(ref = "/pmmlDefinitions.json#/definitions/ResultSet")), description = "PMML result")
+    public Object result(@RequestBody(required = true) Map<String, Object> variables) {
+        return super.result(application, MODEL_NAME, variables);
     }
+
+    @PostMapping(value = "/descriptive", produces = MediaType.APPLICATION_JSON_VALUE, consumes = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (tarilabs) [reply kogito-codegen-modules/kogito-codegen-predictions/src/main/resources/class-templates/PMMLRestResourceSpringTemplate.java:30]: Whatever works for you is fine, I would maybe try to consider this pre-prod if any change is needed 👍

File: drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/AbstractPMMLRestResource.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Map;
+
+import org.kie.api.pmml.PMML4Result;
+import org.kie.kogito.Application;
+
+public abstract class AbstractPMMLRestResource {
+
+
+    protected Object result(Application application, String modelName, Map<String, Object> variables) {
+        org.kie.kogito.prediction.PredictionModel prediction = application.get(org.kie.kogito.prediction.PredictionModels.class).getPredictionModel(modelName);
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/AbstractPMMLRestResource.java:27]: In terms of reusability having "Application" is more useful than PredictionModels - and it is the only reference possible to the "global" environment.

File: kogito-codegen-modules/kogito-codegen-predictions/src/main/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGenerator.java (hunk 6 above)
Comment (gitgabrio) [reply kogito-codegen-modules/kogito-codegen-predictions/src/main/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGenerator.java:195]: @danielezonca 
without the filter, IDE warns about "Optional.get without 'isPresent' check" - 
Maybe is a different expression you are thinking of - that I would be glad to use

**Comment by gitgabrio** [comment]
@danielezonca 
Done modifications

File: kogito-codegen-modules/kogito-codegen-predictions/src/main/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGenerator.java (hunk 6 above)
Comment (danielezonca) [reply kogito-codegen-modules/kogito-codegen-predictions/src/main/java/org/kie/kogito/codegen/prediction/PMMLRestResourceGenerator.java]: You are right, you need to use `flatMap` instead of `map`
```suggestion
        return getAnnotationExpr(source, annotationName)
                .flatMap(annExpr -> getMemberValuePairFromAnnotation(annExpr, CONTENT))
                .flatMap(content -> getMemberValuePairFromMemberValuePair(content, SCHEMA))
                .flatMap(schema -> getNormalAnnotationExprFromMemberValuePair(schema, schemaName))
                .flatMap(schemaAnnotation -> getMemberValuePairFromAnnotation(schemaAnnotation, REF));
```

**Review by jiripetrlik — APPROVED** [review]
Thank you @gitgabrio looks good to me now.

**Comment by gitgabrio** [comment]
Jenkins retest this

**Comment by gitgabrio** [comment]
Jenkins retest this

**Review by danielezonca — APPROVED** [review]
